Assets/Scripts/ConnectionManagement/ConnectionState/HostingState.cs

```diff
@@ -0,0 +1,166 @@
+            var gameReturnStatus = GetConnectStatus(connectionPayload);
+
+            if (gameReturnStatus == ConnectStatus.Success)
+            {
+                SessionManager<SessionPlayerData>.Instance.SetupConnectingPlayerSessionData(clientId, connectionPayload.playerId,
+                    new SessionPlayerData(clientId, connectionPayload.playerName, new NetworkGuid(), 0, true));
+
+                // connection approval will create a player object for you
+                response.Approved = true;
+                response.CreatePlayerObject = true;
+                response.Position = Vector3.zero;
+                response.Rotation = Quaternion.identity;
+                return;
+            }
+
+            // In order for clients to not just get disconnected with no feedback, the server needs to tell the client why it disconnected it.
+            // This could happen after an auth check on a service or because of gameplay reasons (server full, wrong build version, etc)
+            // Since network objects haven't synced yet (still in the approval process), we need to send a custom message to clients, wait for
+            // UTP to update a frame and flush that message, then give our response to NetworkManager's connection approval process, with a denied approval.
+            IEnumerator WaitToDenyApproval()
+            {
+                response.Pending = true; // give some time for server to send connection status message to clients
+                response.Approved = false;
+                ConnectionManager.SendServerToClientSetDisconnectReason(clientId, gameReturnStatus);
+                yield return null; // wait a frame so UTP can flush it's messages on next update
+                response.Pending = false; // connection approval process can be finished.
+            }
+
+            ConnectionManager.SendServerToClientSetDisconnectReason(clientId, gameReturnStatus);
+            m_ConnectionManager.StartCoroutine(WaitToDenyApproval());
+            if (m_LobbyServiceFacade.CurrentUnityLobby != null)
+            {
+                m_LobbyServiceFacade.RemovePlayerFromLobbyAsync(connectionPayload.playerId, m_LobbyServiceFacade.CurrentUnityLobby.Id);
+            }
+        }
+
+        ConnectStatus GetConnectStatus(ConnectionPayload connectionPayload)
+        {
+            if (m_ConnectionManager.NetworkManager.ConnectedClientsIds.Count >= m_ConnectionManager.MaxConnectedPlayers)
+            {
+                return ConnectStatus.ServerFull;
+            }
+
+            if (connectionPayload.isDebug != Debug.isDebugBuild)
+            {
+                return ConnectStatus.IncompatibleBuildType;
+            }
+
+            return SessionManager<SessionPlayerData>.Instance.IsDuplicateConnection(connectionPayload.playerId) ?
+                ConnectStatus.LoggedInAgain : ConnectStatus.Success;
+        }
+
+        IEnumerator WaitToShutdown()
+        {
+            yield return null;
```

Comment:
Just curious:
If the clients disconnect themselves upon receiving the message from the server (do they? still looking around), then would it be over-kill to wait for all of the clients to respond or was that problematic for other reasons (i.e. you want to quickly cancel/end a session and start a new one)?

Reply:
In the end, I thought it made more sense for the clients to not disconnect themselves and wait for Netcode to do it for them when the server shuts down, but the other way around would have worked too.

As for waiting for a response from the clients, the host doesn't really care that all clients receive and handle that custom message, it's just something to let them know that this shut down is voluntary so that they don't try to reconnect to the server. After that, the ball is no longer in its court in my opinion.